pkg/cable/vxlan/vxlan.go

```diff
@@ -0,0 +1,540 @@
+	"github.com/submariner-io/submariner/pkg/cable"
+	"github.com/submariner-io/submariner/pkg/natdiscovery"
+	netlinkAPI "github.com/submariner-io/submariner/pkg/netlink"
+	"github.com/submariner-io/submariner/pkg/types"
+	"github.com/submariner-io/submariner/pkg/util"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog"
+)
+
+const (
+	VxlanIface             = "vxlan-tunnel"
+	VxlanOverhead          = 50
+	VxlanVTepNetworkPrefix = 241
+	CableDriverName        = "vxlan"
+	TableID                = 100
+	defaultNATTPort        = 4500
+)
+
+type Operation int
+
+const (
+	Add Operation = iota
+	Delete
+	Flush
+)
+
+type vxlan struct {
+	localEndpoint types.SubmarinerEndpoint
+	connections   []v1.Connection
+	mutex         sync.Mutex
+	vxlanIface    *vxlanIface
+	netLink       netlinkAPI.Interface
+}
+
+type vxlanIface struct {
+	activeEndpointHostname string
+	link                   *netlink.Vxlan
+	vtepIP                 net.IP
+}
+
+type vxlanAttributes struct {
+	name     string
+	vxlanID  int
+	group    net.IP
+	srcAddr  net.IP
+	vtepPort int
+	mtu      int
+}
+
+func init() {
+	cable.AddDriver(CableDriverName, NewDriver)
+}
+
+func NewDriver(localEndpoint types.SubmarinerEndpoint, localCluster types.SubmarinerCluster) (cable.Driver, error) {
+	var err error
+
+	v := vxlan{
+		localEndpoint: localEndpoint,
+		netLink:       netlinkAPI.New(),
+	}
+
+	port, err := localEndpoint.Spec.GetBackendPort(v1.UDPPortConfig, defaultNATTPort)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the UDP port configuration: %v", err)
+	}
+
+	if err = v.createVxlanInterface(localEndpoint.Spec.Hostname, int(port)); err != nil {
+		return nil, fmt.Errorf("failed to setup Vxlan link: %v", err)
+	}
+
+	return &v, nil
+}
+
+func (v *vxlan) createVxlanInterface(activeEndPoint string, port int) error {
+	ipAddr := v.localEndpoint.Spec.PrivateIP
+
+	vtepIP, err := v.getVxlanVtepIPAddress(ipAddr)
+	if err != nil {
+		return fmt.Errorf("failed to derive the vxlan vtepIP for %s, %v", ipAddr, err)
+	}
+
+	defaultHostIface, err := util.GetDefaultGatewayInterface()
+
+	if err != nil {
+		return errors.Wrapf(err, "Unable to find the default interface on host: %s",
+			v.localEndpoint.Spec.Hostname)
+	}
+
+	// Derive the MTU based on the default outgoing interface
+	vxlanMtu := defaultHostIface.MTU - VxlanOverhead
+
+	attrs := &vxlanAttributes{
+		name:     VxlanIface,
+		vxlanID:  1000,
+		group:    nil,
+		srcAddr:  nil,
+		vtepPort: port,
+		mtu:      vxlanMtu,
+	}
+
+	v.vxlanIface, err = newVxlanIface(attrs, activeEndPoint)
+	if err != nil {
+		return fmt.Errorf("failed to create vxlan interface on Gateway Node: %v", err)
+	}
+
+	v.vxlanIface.vtepIP = vtepIP
+
+	err = v.addIPRule()
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("failed to add ip rule: %v", err)
+	}
+
+	// Enable loose mode (rp_filter=2) reverse path filtering on the vxlan interface.
+	// We won't ever create rp_filter, and its permissions are 644
+	// #nosec G306
```

Comment:
nit - the #nosec comment no longer applies here - in fact I think all the comments here can be removed

Reply:
Done.